pkg/workceptor/kubernetes.go

```diff
@@ -224,6 +237,102 @@ func podRunningAndReady(kw KubeUnit) func(event watch.Event) (bool, error) {
 	return inner
 }
 
+// podInfrastructureSuccess checks if the pod has either successfully started, is pending or is running, or has is successfully terminated.
+// Any other state is considered an infrastructure failure.
+func (ku KubeAPIWrapper) podInfrastructureSuccess(pod *corev1.Pod, containerName string) (bool, string, error) {
+	switch pod.Status.Phase {
+	case corev1.PodFailed:
+		for _, cs := range pod.Status.ContainerStatuses {
```

Comment:
I recommend doing this loop at the top of the function, finding or not finding the container first, before you go into any of this conditional logic, 

All this case, loop, if, case flows are tricky to read